db/db_impl.cc

```diff
@@ -1965,6 +1983,34 @@ WriteBatch* DBImpl::BuildBatchGroup(Writer** last_writer) {
   return result;
 }
 
+
+/**
+ * Test 3 biggest things that cause a thread to block
+ * or delay.  There are other reasons in MakeRoomForWrite()
+ * but those are ignored since penalty and such will reflect them.
+ */
+bool
+DBImpl::RequestNonBlockTicket()
+{
+    bool ret_flag(false);
+
+    // assume ticket succeeds
+    inc_and_fetch(&non_block_tickets_);
+
+    // est_mem_usage_ is last known mem_->ApproximateMemoryUsage(), write_buffer_size is constant
```

Comment:
This is purely a matter of style, but I would find this a bit easier to read if it were a single conditional; something like

```C++
if ( est_mem_usage_ <= options_.write_buffer_size
     && 1==GetThrottleWriteRate()
     && 0==GetLastPenalty() )
   ret_flag = true;
else
   dec_and_fetch(&non_block_tickets_);
```